File: sdks/go/kaivue/cameras.go
package kaivue

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
	"strconv"
)

// CameraService handles camera CRUD operations.
type CameraService struct {
	client *Client
}

// CreateCameraRequest is the input for creating a camera.
type CreateCameraRequest struct {
	Name          string        `json:"name"`
	Description   string        `json:"description,omitempty"`
	IPAddress     string        `json:"ip_address"`
	RecorderID    string        `json:"recorder_id"`
	RecordingMode RecordingMode `json:"recording_mode,omitempty"`
	Labels        []string      `json:"labels,omitempty"`
	Username      string        `json:"username,omitempty"`
	Password      string        `json:"password,omitempty"`
}

// ListCamerasRequest is the input for listing cameras.
type ListCamerasRequest struct {
	Search      string      `json:"search,omitempty"`
	RecorderID  string      `json:"recorder_id,omitempty"`
	StateFilter CameraState `json:"state_filter,omitempty"`
	PageSize    int         `json:"page_size,omitempty"`
	Cursor      string      `json:"cursor,omitempty"`
}

// ListCamerasResponse wraps a page of cameras.
type ListCamerasResponse struct {
	Cameras    []Camera `json:"cameras"`
	NextCursor string   `json:"next_cursor"`
	TotalCount int      `json:"total_count"`
}

// UpdateCameraRequest is the input for updating a camera.
type UpdateCameraRequest struct {
	ID                string        `json:"id"`
	Name              string        `json:"name,omitempty"`
	Description       string        `json:"description,omitempty"`
	RecordingMode     RecordingMode `json:"recording_mode,omitempty"`
	Labels            []string      `json:"labels,omitempty"`
	AudioEnabled      *bool         `json:"audio_enabled,omitempty"`
	MotionSensitivity *int          `json:"motion_sensitivity,omitempty"`
	UpdateMask        []string      `json:"update_mask"`
}

// Create creates a new camera.
func (s *CameraService) Create(ctx context.Context, req *CreateCameraRequest) (*Camera, error) {
	body, err := s.client.post(ctx, "/v1/cameras", req)
	if err != nil {
		return nil, err
	}
	var resp struct {
		Camera Camera `json:"camera"`
	}
	if err := json.Unmarshal(body, &resp); err != nil {
		return nil, fmt.Errorf("kaivue: unmarshal camera: %w", err)
	}
	return &resp.Camera, nil
}

// Get retrieves a camera by ID.
func (s *CameraService) Get(ctx context.Context, id string) (*Camera, error) {
	body, err := s.client.get(ctx, "/v1/cameras/"+id, nil)
	if err != nil {
		return nil, err
	}
	var resp struct {
		Camera Camera `json:"camera"`
	}
	if err := json.Unmarshal(body, &resp); err != nil {
		return nil, fmt.Errorf("kaivue: unmarshal camera: %w", err)
	}
	return &resp.Camera, nil
}

// Update updates a camera.
func (s *CameraService) Update(ctx context.Context, req *UpdateCameraRequest) (*Camera, error) {
	body, err := s.client.patch(ctx, "/v1/cameras/"+req.ID, req)
	if err != nil {
		return nil, err
	}
	var resp struct {
		Camera Camera `json:"camera"`
	}
	if err := json.Unmarshal(body, &resp); err != nil {
		return nil, fmt.Errorf("kaivue: unmarshal camera: %w", err)
	}
	return &resp.Camera, nil
}

// Delete deletes a camera.
func (s *CameraService) Delete(ctx context.Context, id string, purgeRecordings bool) error {
	params := url.Values{}
	if purgeRecordings {
		params.Set("purge_recordings", "true")
	}
	_, err := s.client.delete(ctx, "/v1/cameras/"+id, params)
	return err
}

// List lists cameras with optional filters.
func (s *CameraService) List(ctx context.Context, req *ListCamerasRequest) (*ListCamerasResponse, error) {
	params := url.Values{}
	if req.PageSize > 0 {
		params.Set("page_size", strconv.Itoa(req.PageSize))
	} else {
		params.Set("page_size", "50")
	}
	if req.Search != "" {
		params.Set("search", req.Search)
	}
	if req.RecorderID != "" {
		params.Set("recorder_id", req.RecorderID)
	}
	if req.StateFilter != "" {
		params.Set("state_filter", string(req.StateFilter))
	}
	if req.Cursor != "" {
		params.Set("cursor", req.Cursor)
	}

	body, err := s.client.get(ctx, "/v1/cameras", params)
	if err != nil {
		return nil, err
	}
	var resp ListCamerasResponse
	if err := json.Unmarshal(body, &resp); err != nil {
		return nil, fmt.Errorf("kaivue: unmarshal cameras list: %w", err)
	}
	return &resp, nil
}

// ListAll auto-paginates through all cameras matching the filter.
func (s *CameraService) ListAll(ctx context.Context, req *ListCamerasRequest) ([]Camera, error) {
	var all []Camera
	cursor := req.Cursor
	for {
		r := *req
		r.Cursor = cursor
		resp, err := s.List(ctx, &r)
		if err != nil {
			return nil, err
		}
		all = append(all, resp.Cameras...)
		if resp.NextCursor == "" || len(resp.Cameras) == 0 {
			break
		}
		cursor = resp.NextCursor
	}
	return all, nil
}

File: sdks/python/kaivue/services/cameras.py
"""Camera service client."""

from __future__ import annotations

from typing import Any, Dict, Generator, List, Optional

from kaivue.models.camera import Camera, CameraState, RecordingMode
from kaivue.services.base import BaseService


class CameraService(BaseService):
    """CRUD operations for cameras."""

    def create(
        self,
        name: str,
        ip_address: str,
        recorder_id: str,
        *,
        description: str = "",
        recording_mode: Optional[RecordingMode] = None,
        labels: Optional[List[str]] = None,
        username: str = "",
        password: str = "",
    ) -> Camera:
        body: Dict[str, Any] = {
            "name": name,
            "ip_address": ip_address,
            "recorder_id": recorder_id,
        }
        if description:
            body["description"] = description
        if recording_mode:
            body["recording_mode"] = recording_mode.value
        if labels:
            body["labels"] = labels
        if username:
            body["username"] = username
        if password:
            body["password"] = password

        resp = self._post("/v1/cameras", body)
        return self._parse(Camera, resp["camera"])

    def get(self, camera_id: str) -> Camera:
        resp = self._get(f"/v1/cameras/{camera_id}")
        return self._parse(Camera, resp["camera"])

    def update(
        self,
        camera_id: str,
        *,
        name: Optional[str] = None,
        description: Optional[str] = None,
        recording_mode: Optional[RecordingMode] = None,
        labels: Optional[List[str]] = None,
        audio_enabled: Optional[bool] = None,
        motion_sensitivity: Optional[int] = None,
    ) -> Camera:
        body: Dict[str, Any] = {"id": camera_id}
        mask: List[str] = []
        if name is not None:
            body["name"] = name
            mask.append("name")
        if description is not None:
            body["description"] = description
            mask.append("description")
        if recording_mode is not None:
            body["recording_mode"] = recording_mode.value
            mask.append("recording_mode")
        if labels is not None:
            body["labels"] = labels
            mask.append("labels")
        if audio_enabled is not None:
            body["audio_enabled"] = audio_enabled
            mask.append("audio_enabled")
        if motion_sensitivity is not None:
            body["motion_sensitivity"] = motion_sensitivity
            mask.append("motion_sensitivity")
        body["update_mask"] = mask

        resp = self._patch(f"/v1/cameras/{camera_id}", body)
        return self._parse(Camera, resp["camera"])

    def delete(self, camera_id: str, *, purge_recordings: bool = False) -> None:
        self._request(
            "DELETE",
            f"/v1/cameras/{camera_id}",
            params={"purge_recordings": purge_recordings},
        )

    def list(
        self,
        *,
        search: Optional[str] = None,
        recorder_id: Optional[str] = None,
        state_filter: Optional[CameraState] = None,
        page_size: int = 50,
        cursor: Optional[str] = None,
    ) -> List[Camera]:
        params: Dict[str, Any] = {"page_size": page_size}
        if search:
            params["search"] = search
        if recorder_id:
            params["recorder_id"] = recorder_id
        if state_filter:
            params["state_filter"] = state_filter.value
        if cursor:
            params["cursor"] = cursor

        resp = self._get("/v1/cameras", **params)
        return self._parse_list(Camera, resp.get("cameras", []))

    def list_all(
        self,
        *,
        search: Optional[str] = None,
        recorder_id: Optional[str] = None,
        state_filter: Optional[CameraState] = None,
        page_size: int = 50,
    ) -> Generator[Camera, None, None]:
        """Auto-paginate through all cameras."""
        params: Dict[str, Any] = {}
        if search:
            params["search"] = search
        if recorder_id:
            params["recorder_id"] = recorder_id
        if state_filter:
            params["state_filter"] = state_filter.value
        yield from self._paginate(Camera, "/v1/cameras", "cameras", page_size=page_size, **params)
